CMakeLists.txt

```diff
@@ -405,8 +405,33 @@ if(NOT MSVC)
     )
 endif()
 
-# Python client
+set(UUID_SRC ${PROJECT_SOURCE_DIR}/api/oc_uuid.c)
 if(UNIX)
+    list(APPEND UUID_SRC ${PROJECT_SOURCE_DIR}/port/linux/random.c)
+elseif(WIN32)
+    list(APPEND UUID_SRC ${PROJECT_SOURCE_DIR}/port/windows/random.c)
+endif()
+
+add_library(uuid-shared SHARED ${UUID_SRC})
+set_property(TARGET uuid-shared PROPERTY C_VISIBILITY_PRESET hidden)
+set_property(TARGET uuid-shared PROPERTY VISIBILITY_INLINES_HIDDEN ON)
+target_compile_definitions(uuid-shared PUBLIC OC_SHARED_LIBRARY)
+set_target_properties(uuid-shared PROPERTIES DEFINE_SYMBOL "OC_SHARED_LIBRARY_EXPORTS")
+target_include_directories(uuid-shared PUBLIC
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}>
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/port>
+    $<BUILD_INTERFACE:${PORT_INCLUDE_DIR}>
+    $<INSTALL_INTERFACE:include/iotivity-lite>
+)
+set_target_properties(uuid-shared PROPERTIES
+    OUTPUT_NAME "iotivity-lite-uuid"
+    VERSION ${PROJECT_VERSION}
+    SOVERSION ${PROJECT_VERSION_MAJOR}
+)
+
+# Python client
+if(UNIX OR WIN32)
     set(client-python-lib-obj
```

Comment:
Hmm, I don't think I can reuse the .o files. I need different annotations when I compile for `static` or `shared` lib.

Reply:
@Oxymoron79 I guess the solution here is to create shared version of the object libraries (`common-obj` and `common-obj-shared`, etc.), since they have differing compile definitions. Though they differ only in the compile definitions, so the CMake code for both targets will be very similar. I should probably create a CMake function that creates a library (both static and shared) target to avoid most of the duplicity.

Reply:
Is that really necessary? The point of the object libraries is to generate static and dynamic libraries without the need to recompile the source...

Is your reasoning for this the `declspec(dllexport)` annotation for Windows? If so, I doubt that different object files are needed, because the user of the DLL has to change the annotation to `declspec(dllimport)` when linking to the binary (which was compiled with `declspec(dllexport)`). But my knowledge of Windows development is limited - so take that with a grain of salt :wink:

Reply:
I would like avoid the duplicity if possible, but when I compile `libiotivity-lite-client-python.so.2.2.5` and examine the public symbols using `nm -D` command - I see that all global functions from `common-obj`, `tinycbor-master`, `client-obj` and `python-obj` are public. Thus, adding `C_VISIBILITY` to the final shared library target had no effect. It seems that I have to compile the `.o` files with the macro definitions for a shared library, linking is not enough. However, I haven't tried it the other way around, add definitions for shared library to the object library targets. If static library targets won't break then that could be the way.

(OT: iotivity builds by CYGWIN and MINGW seem to be broken, but I think I've found the issue; fun for Friday afternoon :D )

Reply:
If I understand https://cmake.org/cmake/help/latest/policy/CMP0063.html correctly, you can set this policy to `NEW` and then set the `C_VISIBILITY` on the object library.

Reply:
Thanks, that looks indeed relevant. Though I still think that this code is problematic:

```C
#ifdef OC_SHARED_LIBRARY
#ifdef OC_SHARED_LIBRARY_EXPORT
#define OC_API OC_EXPORT
#else /* !OC_SHARED_LIBRARY_EXPORT*/
#define OC_API OC_IMPORT
#endif /* OC_SHARED_LIBRARY_EXPORT */
#else  /* !OC_SHARED_LIBRARY */
#define OC_API
#endif /* OC_SHARED_LIBRARY */
```

I'm reading about `dllexport` and if I understand it correctly, whatever the compiler does with the attribute is compiled into the `.o` file. I see no way around the fact that the `.o` files are different between static and shared libraries. Shared libraries need `OC_SHARED_LIBRARY` and `OC_SHARED_LIBRARY_EXPORT` definitions during compilation. The only way this works is by compiling the object files with those definitions and hoping that a static library doesn't have a problem with that. 
Though I can think of one side effect, if you compile a static lib from such object files and then link that static lib into a dynamic library, I think the symbols from the static lib will be exposed as global symbols in the final shared library files since they will have the export attribute.

Reply:
> (OT: iotivity builds by CYGWIN and MINGW seem to be broken, but I think I've found the issue; fun for Friday afternoon :D )

I came across this issue as well. One of our Windows developers found that adding this:
```
#ifdef __MINGW32__
#undef interface
#endif
```
in `port/windows/tcpadapter.c` and `port/windows/network_addresses.c`  after the include statements fixes the build.

To me, this seems like quite a hack - so I didn't propose the fix upstream. And I also don't know whether CYGWIN and MINGW build are supported by iotivity-lite.

Reply:
> Thanks, that looks indeed relevant. Though I still think that this code is problematic:
> 
> ```c
> #ifdef OC_SHARED_LIBRARY
> #ifdef OC_SHARED_LIBRARY_EXPORT
> #define OC_API OC_EXPORT
> #else /* !OC_SHARED_LIBRARY_EXPORT*/
> #define OC_API OC_IMPORT
> #endif /* OC_SHARED_LIBRARY_EXPORT */
> #else  /* !OC_SHARED_LIBRARY */
> #define OC_API
> #endif /* OC_SHARED_LIBRARY */
> ```
> 
> I'm reading about `dllexport` and if I understand it correctly, whatever the compiler does with the attribute is compiled into the `.o` file. I see no way around the fact that the `.o` files are different between static and shared libraries. Shared libraries need `OC_SHARED_LIBRARY` and `OC_SHARED_LIBRARY_EXPORT` definitions during compilation. The only way this works is by compiling the object files with those definitions and hoping that a static library doesn't have a problem with that.

I think that it is correct to apply the visibility also on static libraries (btw the current visibility `default` made all symbols globally visible anyways). So, applying the compile definitions to the object library would promote the correct symbol visibility to the static and the shared library, right?

> Though I can think of one side effect, if you compile a static lib from such object files and then link that static lib into a dynamic library, I think the symbols from the static lib will be exposed as global symbols in the final shared library files since they will have the export attribute.

Yes, that's true. I found this question (with an answer) also on stackoverflow: https://stackoverflow.com/questions/2222162/how-to-apply-fvisibility-option-to-symbols-in-static-libraries.

Reply:
Yeah, seems like we should be able to get both static and shared libraries from object on Unixes. Not sure what will happen on MSVC, but I guess if my testing app links and runs it should be alright ...?

Though if the definitions will be added to both static and shared libraries I should change the names: `OC_SHARED_LIBRARY` -> `OC_LIBRARY` and `OC_SHARED_LIBRARY_EXPORT` -> `OC_LIBRARY_EXPORT`

Reply:
> Yeah, seems like we should be able to get both static and shared libraries from object on Unixes. Not sure what will happen on MSVC, but I guess if my testing app links and runs it should be alright ...?

Yes, I think so as well.

> Though if the definitions will be added to both static and shared libraries I should change the names: `OC_SHARED_LIBRARY` -> `OC_LIBRARY` and `OC_SHARED_LIBRARY_EXPORT` -> `OC_LIBRARY_EXPORT`.

Sounds good to me.

Reply:
It seems to work. Compilation of testing apps that use static lib (`uuid-app-static`) and shared lib (`uuid-app-shared`) generated from object lib succeeds on both Linux and Win (see PR jobs). I've also examined `uuid-shared` with `nm -D` on Linux and with `dumpbin /exports` on Win10 locally and the lib only has the manually exposed symbols visible. 

Additional notes:

- CMP0063 is set to NEW by default, so I skipped setting it explicitly
- When compiling a static lib into a shared lib, gcc (at least my 9.3 version) seems to automatically hide all symbols from static library. So one must add compiler flags to make them visible, however this is not an interesting use case currently, so I didn't pursue it further.

OK, so I'll remove the uuid-* testing stuff and we can start annotating public functions and enabling compilation of DLLs on Windows. @WAvdBeek which functions are considered to be public, only those with headers in `include/` folder? Any other? (TBH, in my app that links dynamically with libiotivity-client-server.so, I use functions from all over the place, from `api`, `security`, `port` and `util` folders at least. But I'll work it out.)